pkg/skaffold/build/cache/retrieve.go

```diff
@@ -18,222 +18,131 @@ package cache
 
 import (
 	"context"
-	"fmt"
 	"io"
-	"sync"
+	"strings"
 	"time"
 
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/tag"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/color"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
 	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
-	"github.com/google/go-containerregistry/pkg/name"
 	"github.com/pkg/errors"
-	"github.com/sirupsen/logrus"
 )
 
-var (
-	// For testing
-	hashForArtifact   = getHashForArtifact
-	imgExistsRemotely = imageExistsRemotely
-)
+func (c *cache) Build(ctx context.Context, out io.Writer, tags tag.ImageTags, artifacts []*latest.Artifact, buildAndTest BuildAndTestFn) ([]build.Artifact, error) {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
 
-// ImageDetails holds the Digest and ID of an image
-type ImageDetails struct {
-	Digest string `yaml:"digest,omitempty"`
-	ID     string `yaml:"id,omitempty"`
-}
+	start := time.Now()
 
-// RetrieveCachedArtifacts checks to see if artifacts are cached, and returns tags for cached images, otherwise a list of images to be built
-func (c *Cache) RetrieveCachedArtifacts(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]*latest.Artifact, []build.Artifact, error) {
-	if !c.useCache {
-		return artifacts, nil, nil
+	color.Default.Fprintln(out, "Checking cache...")
+
+	lookup := make(chan []cacheDetails)
+	go func() { lookup <- c.lookupArtifacts(ctx, tags, artifacts) }()
+
+	var results []cacheDetails
+	select {
+	case <-ctx.Done():
+		return nil, context.Canceled
+	case results = <-lookup:
 	}
 
-	start := time.Now()
-	color.Default.Fprintln(out, "Checking cache...")
+	hashByName := make(map[string]string)
+	var needToBuild []*latest.Artifact
+	var alreadyBuilt []build.Artifact
+	for i, artifact := range artifacts {
+		color.Default.Fprintf(out, " - %s: ", artifact.ImageName)
 
-	var wg sync.WaitGroup
-	builtImages := make([]bool, len(artifacts))
-
-	for i := range artifacts {
-		wg.Add(1)
-		i := i
-		go func() {
-			defer wg.Done()
-			ok, err := c.retrieveCachedArtifact(ctx, out, artifacts[i])
-			if ok {
-				builtImages[i] = true
-				return
-			}
-			if err != nil {
-				logrus.Debugf("error finding cached artifact for %s: %v", artifacts[i].ImageName, err)
+		result := results[i]
+		switch result := result.(type) {
+		case failed:
+			return nil, errors.Wrap(result.err, "checking cache")
```

Comment:
In the past, when caching failed, we would log an error and then rebuild the artifact. I'd still expect that as a user, WDYT?